### PR TITLE
JAMES-2292 Add DELETE /mailQueues/{mailQueueName}/mails endpoint

### PR DESCRIPTION
**Pull request by aduprat**

### CODE REVIEW

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java [hunk 1]
```diff
@@ -320,4 +320,159 @@ public void listMailsShouldReturnBadRequestWhenLimitIsInvalid() throws Exception
...
+                .build());
+        String recipient = "recipient@james.org";
+        queue.enQueue(Mails.defaultMail()
+                .recipient(recipient)
+                .build());
+
+        given()
+            .param("sender", sender)
+            .param("name", name)
+            .param("recipient", recipient)
```
Comment (rouazana) [inline server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: so it's an "OR"? Should not it be an "AND"?

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java [hunk 2]
```diff
@@ -238,4 +249,108 @@ private boolean filter(MailQueueItemDTO item, Optional<Boolean> isDelayed) {
...
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .type(ErrorType.INVALID_ARGUMENT)
+                .message("'recipient' should be a mail address (i.e. recipient@james.org)")
+                .cause(e)
+                .haltError();
+        }
+    }
+
+    private Object deleteMails(ManageableMailQueue queue, Optional<MailAddress> maybeSender, Optional<String> maybeName, Optional<MailAddress> maybeRecipient) {
+        if (!maybeSender.isPresent() && !maybeName.isPresent() && !maybeRecipient.isPresent()) {
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]:
```
if (Booleans.count(maybeSender.isPresent(), maybeName.isPresent(), maybeRecipient.isPresent()) == 1)
```

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java [hunk 3]
```diff
@@ -238,4 +249,108 @@ private boolean filter(MailQueueItemDTO item, Optional<Boolean> isDelayed) {
...
+            throw ErrorResponder.builder()
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .type(ErrorType.INVALID_ARGUMENT)
+                .message("At least one of 'sender', 'name' or 'recipient' query parameter should be given.")
+                .haltError();
+        }
+        
+        maybeSender.ifPresent(Throwing.consumer((MailAddress sender) -> queue.remove(Type.Sender, sender.asString())).sneakyThrow());
```
Comment (chibenwa) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]: This, logically should be a `AND`, not a `OR` as you implemented it.

Please reject multiple parameters as needed to match this behavior. CF @mbaechler  comments

Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]: ?

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java (hunk 2 above)
Comment (chibenwa) [reply server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]: ?

**Comment by aduprat** [comment]
> you should add a test with 2 parameters

done

File: server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java [hunk 4]
```diff
@@ -238,4 +250,108 @@ private boolean filter(MailQueueItemDTO item, Optional<Boolean> isDelayed) {
...
+                .statusCode(HttpStatus.BAD_REQUEST_400)
+                .type(ErrorType.INVALID_ARGUMENT)
+                .message("'recipient' should be a mail address (i.e. recipient@james.org)")
+                .cause(e)
+                .haltError();
+        }
+    }
+
+    private Object deleteMails(ManageableMailQueue queue, Optional<MailAddress> maybeSender, Optional<String> maybeName, Optional<MailAddress> maybeRecipient) {
+        if (Booleans.countTrue(maybeSender.isPresent(), maybeName.isPresent(), maybeRecipient.isPresent()) != 1) {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/main/java/org/apache/james/webadmin/routes/MailQueueRoutes.java]: awesome !

File: server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java [hunk 5]
```diff
@@ -320,4 +320,161 @@ public void listMailsShouldReturnBadRequestWhenLimitIsInvalid() throws Exception
...
+        mailQueueFactory.createQueue(FIRST_QUEUE);
+
+        when()
+            .delete(FIRST_QUEUE + "/mails")
+        .then()
+            .statusCode(HttpStatus.BAD_REQUEST_400);
+    }
+
+    @Test
+    public void deleteMailsShouldDeleteMailsWhenSenderIsGiven() throws Exception {
```
Comment (mbaechler) [inline server/protocols/webadmin/webadmin-mailqueue/src/test/java/org/apache/james/webadmin/routes/MailQueueRoutesTest.java]: you should also check that your don't remove too much

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**